### PR TITLE
Darkmode and gif optimization.

### PR DESCRIPTION
**Pull request by 0xM4m4n**
- Added the option to pass minimal or basic for media filter. Originally when browsing the picker the gifs were loading at full resolution. My client browser was loading anywhere from 500kb to 40MB per gif. Using `tnygif` as the query param instead reduces these values by 100-1000x. 
- Added an option to pass light / darkmode theme
- ~Bump to minor version 1.1.0 since this includes some non patch related upgrades. None of these changes should be breaking changes.~
- Tested on storybook locally no issues.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -59,13 +59,15 @@ The following props are accepted by the picker:
...
 | onGifClick | `function` | | Callback function that is called when an gif is clicked. The function receives the [`TenorImage`](#tenorimage) object as a parameter. |
 | autoFocusSearch | `boolean` | `true` | Controls the auto focus of the search input. |
-| contentFilter | `ContentFilter` | `ContentFilter.OFF` | Controls the Tenor [Content filtering](https://developers.google.com/tenor/guides/content-filtering) options. If you are using Typescript you can use `ContentFilter` enum. Possible values are `high`, `medium`, `low`, `off`  |
+| contentFilter | `ContentFilter` | `ContentFilter.OFF` | Controls the Tenor [Content filtering](https://developers.google.com/tenor/guides/
+| mediaFilter | `MediaFilter` | `MediaFilter.BASIC` | Reduce the Number of GIF formats returned in the GIF_OBJECT list. If you are using Typescript you can use `MediaFilter` enum. Possible values are `basic`, `minimal`. For image optimizations it's recommended the picker uses the tinygif or nanogif formats to avoid downloading full gif content to the client while searching. On gif select will return the hires `gif`. Read more [Best practices](https://tenor.com/gifapi/documentation#endpoints-search)  |
```
Comment (MrBartusek) [inline README.md:63]: There is no point in adding a media filter here if it will still be reduced only to use `tinnygif` in `TenorManager`. This option does nothing?

File: package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "gif-picker-react",
   "description": "Tenor Gif Picker for React",
-  "version": "1.0.1",
+  "version": "1.1.0",
```
Comment (MrBartusek) [inline package.json:4]: don't bump versions in PRs

File: src/GifPickerReact.css [hunk 3]
```diff
@@ -1,12 +1,27 @@
...
+    --gpr-highlight-color: #8961a7;
+    --gpr-hover-bg-color: #858688;
+    --gpr-focus-bg-color: #9a9a9b;
+    --gpr-text-color: #f6f6f6;
+    --gpr-search-input-bg-color: #3b3b3b;
+    --gpr-picker-border-color: #2b2b2b;
+    --gpr-bg-color: rgb(42, 41, 41);
+    --gpr-cat-font-color: #f6f6f6;
+}
+
```
Comment (MrBartusek) [inline src/GifPickerReact.css:22]: Yeah you rely on body theme and theme configuration prop is ignored, Look how it's done here https://github.com/ealush/emoji-picker-react/blob/master/src/EmojiPickerReact.css

**Review by MrBartusek — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (0xM4m4n) [reply package.json:4]: fixed

File: README.md (hunk 1 above)
Comment (0xM4m4n) [reply README.md:63]: Yea I can remove it from the options and just use basic in the configuration

File: src/GifPickerReact.css (hunk 3 above)
Comment (0xM4m4n) [reply src/GifPickerReact.css:22]: Ah ok, will update to this instead

Comment (0xM4m4n) [reply src/GifPickerReact.css:22]: This is updated to use similar to EmojiPicker but without clsx or classnames dep package

File: src/GifPickerReact.css [hunk 4]
```diff
@@ -28,7 +43,7 @@
     --gpr-category-list-padding: var(--gpr-horizontal-padding);
     --gpr-category-border-radius: 4px;
     --gpr-category-background-opacity: 40%;
-    --gpr-category-font-color: var(--gpr-picker-border-color);
+    --gpr-category-font-color: var(--gpr-cat-font-color);
```
Comment (MrBartusek) [inline src/GifPickerReact.css:46]: No idea what that does, `--gpr-cat-font-color` is not specified anywhere

File: src/managers/TenorManager.ts [hunk 5]
```diff
@@ -39,7 +38,7 @@ class TenorManager {
 			'key': this.apiKey,
 			'client_key': this.clientKey,
 			'contentfilter': this.contentFilter,
-			'media_filter': MEDIA_FILTER,
+			'media_filter': MediaFilter.BASIC,
```
Comment (MrBartusek) [inline src/managers/TenorManager.ts:41]: This is not a valid media filter

image

https://developers.google.com/tenor/guides/endpoints

File: src/managers/TenorManager.ts [hunk 6]
```diff
@@ -61,7 +60,7 @@ class TenorManager {
 	}
 
 	private praseResult(img: any): TenorImage {
-		const gif = img['media_formats']['gif'];
+		const gif = img['media_formats']['tinygif'];
```
Comment (MrBartusek) [inline src/managers/TenorManager.ts:63]: tinygif can be used for preview and you are returning it to the end user.  You should return the full gif to callback and use these tiny ones for previews in picker.

**Review by MrBartusek — CHANGES_REQUESTED** [review]

**Comment by MrBartusek** [comment]
For the next time. This PR should be actually split into two or three separate ones. One for dark theme, one for tinygif/preview stuff and a third one for other misc changes. That way it would be easier to review.

File: README.md [hunk 7]
```diff
@@ -59,13 +59,14 @@ The following props are accepted by the picker:
 | tenorApiKey | `string` | **Reqired** | Tenor v2 API key, obtained from [Google Cloud Console](https://console.cloud.google.com) |
 | onGifClick | `function` | | Callback function that is called when an gif is clicked. The function receives the [`TenorImage`](#tenorimage) object as a parameter. |
 | autoFocusSearch | `boolean` | `true` | Controls the auto focus of the search input. |
-| contentFilter | `ContentFilter` | `ContentFilter.OFF` | Controls the Tenor [Content filtering](https://developers.google.com/tenor/guides/content-filtering) options. If you are using Typescript you can use `ContentFilter` enum. Possible values are `high`, `medium`, `low`, `off`  |
```
Comment (MrBartusek) [inline README.md:62]: Removed too much of the docs

**Review by MrBartusek — CHANGES_REQUESTED** [review]

File: src/GifPickerReact.tsx [hunk 8]
```diff
@@ -31,7 +32,8 @@ function GifPickerReact(props: GifPickerReactProps): JSX.Element {
 		new TenorManager(settings.tenorApiKey, settings.clientKey,
 			settings.country, settings.locale, settings.contentFilter)
 	), [ ]);
-
+	const newTheme = props.theme === Theme.LIGHT ? 'light' : 'dark';
+	document.body.dataset.theme = newTheme;
```
[the remark this answers is not in the document]
Comment (0xM4m4n) [reply src/GifPickerReact.tsx:36]: This code was removed

File: src/components/header/ClearButton.css [hunk 9]
```diff
@@ -1,8 +1,8 @@
 .GifPickerReact button.gpr-btn-clear-search {
 	position: absolute;
 	right: var(--gpr-search-bar-inner-padding);
-	height: 30px;
-	width: 30px;
+	height: 20px;
+	width: 20px;
```
[the remark this answers is not in the document]
Comment (0xM4m4n) [reply src/components/header/ClearButton.css:5]: Without this change for dark:
image

With the change:
image

File: src/managers/TenorManager.ts (hunk 6 above)
Comment (0xM4m4n) [reply src/managers/TenorManager.ts:63]: No, this is what's causing the browser to load the full gifs since this method is called on search and trending. The gif object returned as all the needed information to get the full resolution gif. The URL is the full resolution and can be used by clients img src or however needed.

File: src/components/PickerMain.tsx [hunk 10]
```diff
@@ -1,20 +1,20 @@
...
 
 function PickerMain({ children }: PickerMainProps): JSX.Element {
 	const settings = useContext(SettingsContext);
-
 	const style: React.CSSProperties = {
 		height: settings.height,
 		width: settings.width
 	};
-
+	const pickerClass = 'GifPickerReact gpr-main' + (settings.theme === Theme.DARK ? ' gpr-dark-theme' : '');
```
Comment (0xM4m4n) [inline src/components/PickerMain.tsx:15]: using this string to avoid dependecy on classnames or clsx package

File: src/managers/TenorManager.ts [hunk 11]
```diff
@@ -39,7 +38,7 @@ class TenorManager {
 			'key': this.apiKey,
 			'client_key': this.clientKey,
 			'contentfilter': this.contentFilter,
-			'media_filter': MEDIA_FILTER,
+			'media_filter': MediaFilter.TINYGIF,
```
Comment (0xM4m4n) [inline src/managers/TenorManager.ts:41]: surprisingly enough it works for both minimal and basic however you need to only parse the tinygif result to reduce load on clients. The metadata associated to the obj result has enought information to generate the fullsized version

File: src/managers/TenorManager.ts (hunk 6 above)
Comment (MrBartusek) [reply src/managers/TenorManager.ts]: this is the only place where `TenorImage` objects are generated and then passed to `onGifClick`

File: src/GifPickerReact.css (hunk 4 above)
Comment (0xM4m4n) [reply src/GifPickerReact.css]: Updated was a workaround

File: src/managers/TenorManager.ts (hunk 5 above)
Comment (0xM4m4n) [reply src/managers/TenorManager.ts]: The api accepts basic and minimal this and returns the different formats.

File: src/managers/TenorManager.ts (hunk 6 above)
Comment (0xM4m4n) [reply src/managers/TenorManager.ts]: current version clicking trending: image

using tinygif:
image

Comment (MrBartusek) [reply src/managers/TenorManager.ts]: image

image

File: src/components/PickerMain.tsx (hunk 10 above)
Comment (MrBartusek) [reply src/components/PickerMain.tsx:15]: move it inline to classname with template literals

File: src/managers/TenorManager.ts (hunk 11 above)
Comment (MrBartusek) [reply src/managers/TenorManager.ts:41]: > The metadata associated to the obj result has enought information to generate the fullsized version

it is not

File: src/managers/TenorManager.ts (hunk 5 above)
Comment (MrBartusek) [reply src/managers/TenorManager.ts]: Using undocumented api is not good idea

File: src/GifPickerReact.css [hunk 12]
```diff
@@ -5,8 +5,8 @@
     --gpr-text-color: #858585;
     --gpr-search-input-bg-color: #f6f6f6;
     --gpr-picker-border-color: #e7e7e7;
+    --gpr-body-category-font-color: #e7e7e7;
```
Comment (MrBartusek) [inline src/GifPickerReact.css:8]: --gpr-category-font-color = --gpr-body-category-font-color = --gpr-picker-border-color there is no point in another redundant varable

File: src/managers/TenorManager.ts (hunk 5 above)
Comment (0xM4m4n) [reply src/managers/TenorManager.ts]: It's still documented here: https://tenor.com/gifapi/documentation#endpoints-search

Comment (0xM4m4n) [reply src/managers/TenorManager.ts]: Updated.

File: src/managers/TenorManager.ts (hunk 6 above)
Comment (0xM4m4n) [reply src/managers/TenorManager.ts]: Updated to use fullgif dimensions. You can see that when specifying to use the tiny gif the netwrok request size in bytes is drastically reduced. Using the full gif size will give the callback the correct dimensions

File: src/GifPickerReact.css (hunk 12 above)
Comment (0xM4m4n) [reply src/GifPickerReact.css:8]: removed

**Comment by 0xM4m4n** [comment]
> 

Agreed, apologies for all in one, was working on it late last night and got carried away

File: src/managers/TenorManager.ts (hunk 11 above)
Comment (0xM4m4n) [reply src/managers/TenorManager.ts]: updated to use gif dimensions and not tiny gif

File: src/managers/TenorManager.ts [hunk 13]
```diff
@@ -61,15 +60,18 @@ class TenorManager {
 	}
 
 	private praseResult(img: any): TenorImage {
-		const gif = img['media_formats']['gif'];
+		//Use tinygif object for compnent render. Tenor urls are the same betweent tinygif and gif
+		const tinyGif = img['media_formats'][MediaFilter.TINYGIF];
```
Comment (0xM4m4n) [inline src/managers/TenorManager.ts:64]: if you remove this line and just use the next line, the media loads are in the MBs.

File: src/GifPickerReact.css [hunk 14]
```diff
@@ -28,7 +27,7 @@
     --gpr-category-list-padding: var(--gpr-horizontal-padding);
     --gpr-category-border-radius: 4px;
     --gpr-category-background-opacity: 40%;
-    --gpr-category-font-color: var(--gpr-picker-border-color);
+    --gpr-category-font-color: #e7e7e7;
```
Comment (MrBartusek) [inline src/GifPickerReact.css:30]: Still useless change, just rever it

Comment (MrBartusek) [inline src/GifPickerReact.css:30]: Still useless change, just rever it

File: src/GifPickerReact.tsx [hunk 15]
```diff
@@ -31,7 +32,6 @@ function GifPickerReact(props: GifPickerReactProps): JSX.Element {
 		new TenorManager(settings.tenorApiKey, settings.clientKey,
 			settings.country, settings.locale, settings.contentFilter)
 	), [ ]);
-
```
Comment (MrBartusek) [inline src/GifPickerReact.tsx:34]: unnecessary whitespace change

File: src/components/PickerMain.tsx [hunk 16]
```diff
@@ -1,20 +1,19 @@
...
 function PickerMain({ children }: PickerMainProps): JSX.Element {
 	const settings = useContext(SettingsContext);
-
 	const style: React.CSSProperties = {
 		height: settings.height,
 		width: settings.width
 	};
-
 	return (
-		<aside className='GifPickerReact gpr-main' style={style}>
```
Comment (MrBartusek) [inline src/components/PickerMain.tsx:17]: unnecessary whitespace change

File: src/managers/TenorManager.ts (hunk 13 above)
Comment (MrBartusek) [reply src/managers/TenorManager.ts:64]: Then save tinygifs in property named `preview` and use it for displaying previews in picker but still pass regular gif to callback

File: src/types/exposedTypes.ts [hunk 17]
```diff
@@ -10,6 +10,13 @@ export enum Theme {
     AUTO = 'dark'
 }
 
+export enum MediaFilter {
+    TINYGIF = 'tinygif',
+    GIF = 'gif',
+    TINYMP4 = 'tinymp4',
+    MP4 = 'mp4'
+}
+
```
Comment (MrBartusek) [inline src/types/exposedTypes.ts:19]: Unused enum

**Review by MrBartusek — CHANGES_REQUESTED** [review]

File: src/hooks/useSettings.ts [hunk 18]
```diff
@@ -32,7 +32,8 @@ function useSettings(props: GifPickerReactProps): GifPickerSettings {
 		contentFilter: props.contentFilter ?? ContentFilter.OFF,
 		height: praseDimension(props.height ?? 450),
 		width: praseDimension(props.width ?? 350),
-		categoryHeight: praseDimension(props.categoryHeight ?? 100)
+		categoryHeight: praseDimension(props.categoryHeight ?? 100),
+		theme: props.theme ?? Theme.LIGHT
```
Comment (MrBartusek) [inline src/hooks/useSettings.ts:36]: You should handle also the auto theme here

**Review by MrBartusek — CHANGES_REQUESTED** [review]

File: src/managers/TenorManager.ts (hunk 13 above)
Comment (0xM4m4n) [reply src/managers/TenorManager.ts]: I believe my latest update accomplishes this.

File: src/GifPickerReact.css (hunk 14 above)
Comment (0xM4m4n) [reply src/GifPickerReact.css:30]: This is needed for DarkMode otherwise the categories are hard to read because it was relying on border color before.

Comment (0xM4m4n) [reply src/GifPickerReact.css:30]: removed:
image

changed:
image

**Review by MrBartusek — CHANGES_REQUESTED** [review]
This is probably close to merging. You can also add ability to select dark theme in storybook for easier testing https://github.com/MrBartusek/gif-picker-react/blob/master/src/stories/GifPicker.stories.tsx#L10-L35

**Comment by MrBartusek** [comment]
image
Close hover is too light

**Comment by MrBartusek** [comment]
image
Loading placeholders are white

File: README.md [hunk 19]
```diff
@@ -95,7 +107,12 @@ This is an example `TenorImage` object:
   tags: [ "American Psycho", "Patrick Bateman", "American", "psycho"],
   url: "https://media.tenor.com/5lLcKZgmIhgAAAAC/american-psycho-patrick-bateman.gif",
   height: 240,
-  width: 244
+  width: 244,
+  preview: {
+    url: "https://media.tenor.com/5lLcKZgmIhgAAAAC/american-psycho-patrick-bateman.gif",
```
Comment (MrBartusek) [inline README.md:112]: `https://media.tenor.com/5lLcKZgmIhgAAAAM/american-psycho-patrick-bateman.gif`

Comment (0xM4m4n) [reply README.md:112]: Wow that ONE letter difference I never noticed 😅 makes a lot more sense now haha

File: src/GifPickerReact.css [hunk 20]
```diff
@@ -39,6 +38,16 @@
...
 
+.GifPickerReact.gpr-dark-theme {
+    --gpr-highlight-color: #8961a7;
+    --gpr-hover-bg-color: #858688;
+    --gpr-focus-bg-color: #9a9a9b;
+    --gpr-text-color: #f6f6f6;
+    --gpr-search-input-bg-color: #3b3b3b;
+    --gpr-picker-border-color: #2b2b2b;
+    --gpr-bg-color: rgb(42, 41, 41);
+}
```
[the remark this answers is not in the document]
Comment (0xM4m4n) [reply src/GifPickerReact.css:49]: Ok Ill follow them, I kinda liked the flair a bit different lol

[the remark this answers is not in the document]
Comment (0xM4m4n) [reply src/GifPickerReact.css]: Side note, how do I override these in my own app? I couldnt seem to get the CSS to be applied no matter what I tried. Like if I want to change the hover border color for example for both light and dark mode on my app? Its a React webapp with some scss styles

**Review by MrBartusek — APPROVED** [review]

**Comment by MrBartusek** [comment]
Thank you!

**Comment by 0xM4m4n** [comment]
Sorry missed those last few comments the other day. Thanks for making the changes and merging!
